hledger-api/hledger-api.1

.TH "hledger\-api" "1" "March 2019" "hledger\-api 1.14" "hledger User Manuals"



.SH NAME
.PP
hledger\-api \- web API server for the hledger accounting tool
.SH SYNOPSIS
.PP
\f[C]hledger\-api\ [OPTIONS]\f[]
.PD 0
.P
.PD
\f[C]hledger\ api\ \-\-\ [OPTIONS]\f[]
.SH DESCRIPTION
.PP
hledger is a cross\-platform program for tracking money, time, or any
other commodity, using double\-entry accounting and a simple, editable
file format.
hledger is inspired by and largely compatible with ledger(1).
.PP
hledger\-api is a simple web API server, intended to support
client\-side web apps operating on hledger data.
It comes with a series of simple client\-side app examples, which drive
its evolution.
.PP
Like hledger, it reads data from one or more files in hledger journal,
timeclock, timedot, or CSV format specified with \f[C]\-f\f[], or
\f[C]$LEDGER_FILE\f[], or \f[C]$HOME/.hledger.journal\f[] (on windows,
perhaps \f[C]C:/Users/USER/.hledger.journal\f[]).
For more about this see hledger(1), hledger_journal(5) etc.
.PP
The server listens on IP address 127.0.0.1, accessible only to local
requests, by default.
You can change this with \f[C]\-\-host\f[], eg
\f[C]\-\-host\ 0.0.0.0\f[] to listen on all addresses.
Note there is no other access control, and hledger\-api allows file
browsing, so on shared machines you will certainly need to put it behind
an authenticating proxy to restrict access.
.PP
You can change the TCP port it listens on (default: 8001) with
\f[C]\-p\ PORT\f[].
.PP
API methods look like:
.IP
.nf
\f[C]
/api/v1/accountnames
/api/v1/transactions
/api/v1/prices
/api/v1/commodities
/api/v1/accounts
/api/v1/accounts/ACCTNAME
\f[]
.fi
.PP
See \f[C]/api/swagger.json\f[] for a full list in Swagger 2.0 format.
(Or you can run \f[C]hledger\-api\ \-\-swagger\f[] to print this in the
console.)
.PP
hledger\-api also serves files, from the current directory by default,
and the \f[C]/\f[] path will also show a directory listing.
This is convenient for serving client\-side web code, in addition to the
server\-side api.
.SH OPTIONS
.PP
Note: if invoking hledger\-api as a hledger subcommand, write
\f[C]\-\-\f[] before options as shown above.
.TP
.B \f[C]\-f\ \-\-file=FILE\f[]
use a different input file.
For stdin, use \- (default: \f[C]$LEDGER_FILE\f[] or
\f[C]$HOME/.hledger.journal\f[])
.RS
.RE
.TP
.B \f[C]\-d\ \-\-static\-dir=DIR\f[]
serve files from a different directory (default: \f[C]\&.\f[])
.RS
.RE
.TP
.B \f[C]\-\-host=IPADDR\f[]
listen on this IP address (default: 127.0.0.1)
.RS
.RE
.TP
.B \f[C]\-p\ \-\-port=PORT\f[]
listen on this TCP port (default: 8001)
.RS
.RE
.TP
.B \f[C]\-\-swagger\f[]
print API docs in Swagger 2.0 format, and exit
.RS
.RE
.TP
.B \f[C]\-\-version\f[]
show version
.RS
.RE
.TP
.B \f[C]\-h\ \-\-help\f[]
show usage
.RS
.RE
.SH ENVIRONMENT
.PP
\f[B]LEDGER_FILE\f[] The journal file path when not specified with
\f[C]\-f\f[].
Default: \f[C]~/.hledger.journal\f[] (on windows, perhaps
\f[C]C:/Users/USER/.hledger.journal\f[]).
.SH FILES
.PP
Reads data from one or more files in hledger journal, timeclock,
timedot, or CSV format specified with \f[C]\-f\f[], or
\f[C]$LEDGER_FILE\f[], or \f[C]$HOME/.hledger.journal\f[] (on windows,
perhaps \f[C]C:/Users/USER/.hledger.journal\f[]).
.SH BUGS
.PP
The need to precede options with \f[C]\-\-\f[] when invoked from hledger
is awkward.


.SH "REPORTING BUGS"
Report bugs at http://bugs.hledger.org
(or on the #hledger IRC channel or hledger mail list)

.SH AUTHORS
Simon Michael <simon@joyful.com> and contributors

.SH COPYRIGHT

Copyright (C) 2007-2016 Simon Michael.
.br
Released under GNU GPL v3 or later.

.SH SEE ALSO
hledger(1), hledger\-ui(1), hledger\-web(1), hledger\-api(1),
hledger_csv(5), hledger_journal(5), hledger_timeclock(5), hledger_timedot(5),
ledger(1)

http://hledger.org


NEWS
hledger news
============

2010/05/23 hledger 0.10
........................

  * fix too-loose testpack dependency, missing safe dependency
  * fix ghc 6.12 compatibility with -fweb
  * fix handling of non-ascii arguments with ghc 6.12
  * fix "0.8" in --version output
  * fix an occasional stack overflow error due to infinite recursion in Posting/Transaction equality tests
  * the -fwebhappstack build flag is gone for now, to avoid a cabal problem
  * parsing: if there is no description, don't require a space after the transaction date
  * parsing: balance balanced-virtual postings separately, allow them to have an implicit amount
  * parsing: timelog entries now generate balanced transactions, using virtual postings
  * parsing: simpler high-level parse error message
  * parsing: clearer bad date errors
  * add: fix wrongful program exit on bad dates
  * print: negative account patterns now exclude transactions containing any posting to a matched account
  * vty: rename the ui command to vty for consistency
  * vty: fix restricted account scope when backing up to top level
  * web: fix non-ascii handling with ghc 6.12
  * web: fix a bug possibly affecting reload-on-change
  * consolidate module namespace under Hledger, api cleanups

  Stats:
  44 days, 81 commits since last release.
  Now at 4904 lines of code including tests, 144 tests, 53% coverage.

2010/04/10 hledger 0.9
......................

  * ghc 6.12 support
  * split off hledger-lib package, containing core types & utils
  * parsing: ignore D, C, N, tag, end tag directives; we should now accept any ledger 2.6 file
  * parsing: allow numbers in commodities if double-quoted, like ledger
  * parsing: allow transactions with empty descriptions
  * parsing: show a better error for illegal month/day numbers in dates
  * parsing: don't ignore trailing junk in a smart date, eg in web add form
  * parsing: don't ignore unparsed text following an amount
  * parsing: @ was being treated as a currency symbol
  * add: fix precision handling in default amounts (#19)
  * add: elide last amount in added transactions
  * convert: keep original description by default, allow backreferences in replace pattern
  * convert: basic csv file checking, warn instead of dying when it looks wrong
  * convert: allow blank/comment lines at end of rules file
  * print: always show zero amounts as 0, hiding any commodity/decimal places/price, like ledger
  * register: fix bad layout with years < 1000
  * register: fix a Prelude.head error with reporting interval, --empty, and --depth
  * register: fix a regression, register should not show posting comments
  * register: with --empty, intervals should continue to ends of the specified period
  * stats: better output when last transaction is in the future
  * stats: show commodity symbols, account tree depth, reorder slightly
  * web: -fweb now builds with simpleserver; to get happstack, use -fwebhappstack instead
  * web: pre-fill the add form with today's date
  * web: help links, better search form wording
  * web: show a proper error for a bad date in add form (#17)
  * web: fix for unicode search form values
  * web: fix stack overflow caused by regexpr, and handle requests faster (#14)
  * web: look for more-generic browser executables
  * web: more robust browser starting (#6)
  * error message cleanups
  * more tests, refactoring, docs

  Stats:
  58 days, 2 contributors, 102 commits since last release.
  Now at 3983 lines of non-test code, 139 tests, 53% coverage.

2010/02/11 hledger 0.8
......................

  * parsing: in date=date2, use first date's year as a default for the second
  * add: ctrl-d doesn't work on windows, suggest ctrl-c instead
  * add: --no-new-accounts option disallows new accounts (Roman Cheplyaka)
  * add: re-use the previous transaction's date as default (Roman Cheplyaka)
  * add: a command-line argument now filters by account during history matching (Roman Cheplyaka)
  * chart: new command, generates balances pie chart (requires -fchart flag, gtk2hs) (Roman Cheplyaka, Simon Michael)
  * register: make reporting intervals honour a display expression (#18)
  * web: fix help link
  * web: use today as default when adding with a blank date
  * web: re-enable account/period fields, they seem to be fixed, along with file re-reading (#16)
  * web: get static files from the cabal data dir, or the current dir when using make (#13)
  * web: preserve encoding during add, assuming it's utf-8 (#15)
  * fix some non-utf8-aware file handling (#15)
  * filter ledger again for each command, not just once at program start
  * refactoring, clearer data types

  Stats:
  62 days, 2 contributors, 76 commits since last release.
  Now at 3464 lines of non-test code, 97 tests, 53% test coverage.

2009/12/11 hledger 0.7
........................

  * price history support (first cut):
    P directives now work, though differently from c++ ledger. Each
    posting amount takes its fixed unit price from the price history (or
    @) when available. This is simple and useful for things like foreign
    currency expenses (but not investment tracking). Like ledger, balance
    and register don't show amount prices any more, and don't separate
    differently-priced amounts. Unlike ledger, print shows all amount
    prices, and supports -B.
  * --effective option, will use transactions' effective dates if any
  * convert: new rules file format, find/create rules file automatically,
    more robust parsing, more useful --debug output
  * print: always sort by date, fix long account name truncation, align
    amounts, show end of line comments, show all amounts for clarity
    (don't elide the final balancing amount)
  * ui: use vty 4, fixes non-ascii and gnome terminal problems (issues #3, #4)
  * web: allow data entry, react to data file changes, better layout, help
    links, remove histogram command and filter fields for now, fix bad

hledger-api/hledger-api.hs
{-# LANGUAGE CPP                 #-}
{-# LANGUAGE DataKinds           #-}
{-# LANGUAGE DeriveGeneric       #-}
{-# LANGUAGE OverloadedStrings   #-}
{-# LANGUAGE PolyKinds           #-}
{-# LANGUAGE QuasiQuotes         #-}
{-# LANGUAGE TypeFamilies        #-}
{-# LANGUAGE TypeOperators       #-}
{-# LANGUAGE FlexibleInstances   #-}
{-# LANGUAGE RecordWildCards     #-}
{-# LANGUAGE ScopedTypeVariables #-}

module Main where

import           Lens.Micro ((&), (.~))
import           Control.Monad
import           Data.Aeson
import qualified Data.ByteString.Lazy.Char8 as BL8
import           Data.Decimal
import           Data.Default
import qualified Data.Map as M
import           Data.Proxy
import           Data.String (fromString)
import           Data.Swagger
import           Data.Text hiding (map,reverse)
import           Network.Wai as Wai
import           Network.Wai.Handler.Warp as Warp
import           Network.Wai.Middleware.RequestLogger
import           Safe
import           Servant
import           Servant.Swagger
import           System.Console.Docopt
import           System.Environment (getArgs)
import           System.Exit
import           Text.Printf

import Hledger.Query
import Hledger.Cli hiding (Reader, version)

#ifdef VERSION
hledgerApiVersion = VERSION
#else
hledgerApiVersion = "dev build"
#endif

-- https://github.com/docopt/docopt.hs#readme
-- XXX VERSION is "X.Y" and the quotes appear in the output
doc :: Docopt
doc = [docopt|
hledger-api VERSION

Serves hledger data and reports as a JSON web API.

Usage:
  hledger-api [options]
    start API server
  hledger-api --swagger
    print API docs in Swagger 2.0 format
  hledger-api --version
  hledger-api -h|--help

Options:
  -f --file FILE   use a different input file
                   (default: $LEDGER_FILE or ~/.hledger.journal)
  -d --static-dir  DIR  serve files from a different directory
                   (default: .)
     --host IPADDR listen on this IP address (default: 127.0.0.1)
  -p --port PORT   listen on this TCP port (default: 8001)
     --version     show version
  -h --help        show usage
|]

swaggerSpec :: Swagger
swaggerSpec = toSwagger (Proxy :: Proxy HledgerApi)
  & info.title       .~ "hledger API"
  & info.version     .~ pack hledgerApiVersion
  & info.description .~ Just "This is the API provided by hledger-api for reading hledger data"
  & info.license     .~ Just (License "GPLv3+" (Nothing))

main :: IO ()
main = do
  args <- getArgs >>= parseArgsOrExit doc
  when (isPresent args (shortOption 'h') || isPresent args (longOption "help")) $ exitWithUsage doc
  when (isPresent args (longOption "version")) $ putStrLn ("hledger-api " ++ hledgerApiVersion) >> exitSuccess
  when (isPresent args (longOption "swagger")) $ BL8.putStrLn (encode swaggerSpec) >> exitSuccess
  let
    defh = "127.0.0.1"
    h = getArgWithDefault args defh (longOption "host")
    defp = "8001"
  p <- case readMay $ getArgWithDefault args defp (longOption "port") of
        Nothing -> exitWithUsage doc
        Just n  -> return n
  deff <- defaultJournalPath
  let f = getArgWithDefault args deff (longOption "file")
  requireJournalFileExists f
  let
    defd = "."
    d = getArgWithDefault args defd (longOption "static-dir")
  readJournalFile def f >>= either error' (serveApi h p d f)

serveApi :: String -> Int -> FilePath -> FilePath -> Journal -> IO ()
serveApi h p d f j = do
  printf "Starting web api http://%s:%d/api/v1 for %s\n" h p f
  printf "and file server  http://%s:%d        for %s/\n" h p d
  printf "Press ctrl-c to quit\n"
  let warpsettings = defaultSettings
        & setHost (fromString h)
        & setPort p
  Warp.runSettings warpsettings $
    logStdout $
    hledgerApiApp d j

type HledgerApi =
  "api" :> "v1" :>
    (
         "accountnames" :> Get '[JSON] [AccountName]
    :<|> "transactions" :> Get '[JSON] [Transaction]
    :<|> "prices"       :> Get '[JSON] [MarketPrice]
    :<|> "commodities"  :> Get '[JSON] [CommoditySymbol]
    :<|> "accounts"     :> Get '[JSON] [Account]
    :<|> "accounts"     :> Capture "acct" AccountName :> Get '[JSON] AccountTransactionsReport
    )

type HledgerSwaggerApi =
       "swagger.json" :> Get '[JSON] Swagger
  :<|> HledgerApi

type HledgerSwaggerFilesApi =
       HledgerSwaggerApi
  :<|> Raw

hledgerApiApp :: FilePath -> Journal -> Wai.Application
hledgerApiApp staticdir j = Servant.serve api server
  where
    api :: Proxy HledgerSwaggerFilesApi
    api = Proxy

    server :: Server HledgerSwaggerFilesApi
    server =
      (
           return swaggerSpec
      --
      :<|> accountnamesH
      :<|> transactionsH
      :<|> pricesH
      :<|> commoditiesH
      :<|> accountsH
      :<|> accounttransactionsH
      )
      --
      :<|> serveDirectoryFileServer staticdir
      where
        accountnamesH = return $ journalAccountNames j
        transactionsH = return $ jtxns j
        pricesH       = return $ jmarketprices j
        commoditiesH  = return $ (M.keys . jinferredcommodities) j
        accountsH     = return $ ledgerTopAccounts $ ledgerFromJournal Hledger.Query.Any j
        accounttransactionsH (a::AccountName) = do
          -- d <- liftIO getCurrentDay
          let
            ropts = defreportopts
            -- ropts' = ropts {depth_=Nothing
            --                ,balancetype_=HistoricalBalance
            --                }
            q = Hledger.Query.Any --filterQuery (not . queryIsDepth) $ queryFromOpts d ropts'
            thisacctq = Acct $ accountNameToAccountRegex a -- includes subs
          return $ accountTransactionsReport ropts j q thisacctq

-- avoiding https://github.com/bos/aeson/issues/290 - no longer needed ?
--instance ToJSON Status where toJSON = genericToJSON defaultOptions -- avoiding https://github.com/bos/aeson/issues/290
--instance ToJSON GenericSourcePos where toJSON = genericToJSON defaultOptions
--instance ToJSON Decimal where toJSON = toJSON . show
--instance ToJSON Amount where toJSON = genericToJSON defaultOptions
--instance ToJSON AmountStyle where toJSON = genericToJSON defaultOptions
--instance ToJSON Side where toJSON = genericToJSON defaultOptions
--instance ToJSON DigitGroupStyle where toJSON = genericToJSON defaultOptions
--instance ToJSON MixedAmount where toJSON = genericToJSON defaultOptions
--instance ToJSON BalanceAssertion where toJSON = genericToJSON defaultOptions
--instance ToJSON Price where toJSON = genericToJSON defaultOptions
--instance ToJSON MarketPrice where toJSON = genericToJSON defaultOptions
--instance ToJSON PostingType where toJSON = genericToJSON defaultOptions
--instance ToJSON Posting where
--  toJSON Posting{..} =
--    object
--    ["pdate"             .= toJSON pdate
--    ,"pdate2"            .= toJSON pdate2
--    ,"pstatus"           .= toJSON pstatus
--    ,"paccount"          .= toJSON paccount
--    ,"pamount"           .= toJSON pamount
--    ,"pcomment"          .= toJSON pcomment
--    ,"ptype"             .= toJSON ptype
--    ,"ptags"             .= toJSON ptags
--    ,"pbalanceassertion" .= toJSON pbalanceassertion
--    ,"ptransactionidx"   .= toJSON (maybe "" (show.tindex) ptransaction)
--    ]
--instance ToJSON Transaction where toJSON = genericToJSON defaultOptions
--instance ToJSON Account where
--  toJSON a =
--    object
--    ["aname"        .= toJSON (aname a)
--    ,"aebalance"    .= toJSON (aebalance a)
--    ,"aibalance"    .= toJSON (aibalance a)
--    ,"anumpostings" .= toJSON (anumpostings a)
--    ,"aboring"      .= toJSON (aboring a)
--    ,"aparentname"  .= toJSON (maybe "" aname $ aparent a)
--    ,"asubs"        .= toJSON (map toJSON $ asubs a)
--    ]

-- Convert things to JSON for serving to clients
instance ToJSON Status
instance ToJSON GenericSourcePos
instance ToJSON Decimal where toJSON = toJSON . show
instance ToJSON Amount
instance ToJSON AmountStyle
instance ToJSON Side
instance ToJSON DigitGroupStyle
instance ToJSON MixedAmount
instance ToJSON BalanceAssertion
instance ToJSON Price
instance ToJSON MarketPrice
instance ToJSON PostingType
instance ToJSON Posting where
  toJSON Posting{..} =
    object
    ["pdate"             .= toJSON pdate
    ,"pdate2"            .= toJSON pdate2
    ,"pstatus"           .= toJSON pstatus
    ,"paccount"          .= toJSON paccount
    ,"pamount"           .= toJSON pamount
    ,"pcomment"          .= toJSON pcomment
    ,"ptype"             .= toJSON ptype
    ,"ptags"             .= toJSON ptags
    ,"pbalanceassertion" .= toJSON pbalanceassertion
    ,"ptransactionidx"   .= toJSON (maybe "" (show.tindex) ptransaction)
    ]
instance ToJSON Transaction
instance ToJSON Account where
  toJSON a =
    object
    ["aname"        .= toJSON (aname a)
    ,"aebalance"    .= toJSON (aebalance a)
    ,"aibalance"    .= toJSON (aibalance a)
    ,"anumpostings" .= toJSON (anumpostings a)
    ,"aboring"      .= toJSON (aboring a)
    ,"aparentname"  .= toJSON (maybe "" aname $ aparent a)
    ,"asubs"        .= toJSON (map toJSON $ asubs a)
    ]

-- convert things to Schema for swagger API description
instance ToSchema Status
instance ToSchema GenericSourcePos
instance ToSchema Decimal
 where
  declareNamedSchema _proxy = pure $ NamedSchema (Just "Decimal") schema
   where
     schema = mempty
       & type_   .~ SwaggerNumber
       & example .~ Just (toJSON (100 :: Decimal))
instance ToSchema Amount
instance ToSchema AmountStyle
instance ToSchema Side
instance ToSchema DigitGroupStyle
instance ToSchema MixedAmount
instance ToSchema BalanceAssertion
instance ToSchema Price
#if MIN_VERSION_swagger2(2,1,5)
  where declareNamedSchema = genericDeclareNamedSchemaUnrestricted defaultSchemaOptions
#endif
instance ToSchema MarketPrice
instance ToSchema PostingType
instance ToSchema Posting
instance ToSchema Transaction
instance ToSchema Account
instance ToSchema AccountDeclarationInfo
-- instance ToSchema AccountTransactionsReport
